sdk/src/main/java/com/google/cloud/dataflow/sdk/util/common/Counter.java

```diff
@@ -418,17 +434,160 @@ public boolean isCompatibleWith(Counter<?> that) {
 
   //////////////////////////////////////////////////////////////////////////////
 
-  /** The name of this counter. */
-  protected final String name;
-
   /** The kind of aggregation function to apply to this counter. */
   protected final AggregationKind kind;
 
   /** The commit state of this counter. **/
   protected final AtomicReference<CommitState> commitState;
 
+  /** A hashable name structure which will uniquely identify a counter. */
+  public interface Name {
+    public String getName();
+    public NameContext getContext();
+  }
+
+  /** A StructuredName which uses the Original step name from the NameContext. */
+  public static class StructuredNameOriginal implements Name {
+
+    /** The name of this counter. */
+    private final String name;
+
+    /** The context this counter was named in.  */
+    private final NameContext context;
+
+    public StructuredNameOriginal(String name, NameContext context) {
+      this.name = name;
+      this.context = checkNotNull(context);
+    }
+
+    @Override
+    public String getName() {
+      return this.name;
+    }
+
+    @Override
+    public NameContext getContext() {
+      return this.context;
+    }
+
+    @Override
+    public String toString() {
+      StringBuilder sb = new StringBuilder();
+      sb.append("Original name: " + name + ", ");
+      sb.append("context: " + context.toString() + ", ");
+      return sb.toString();
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(name, context.originalName());
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (other instanceof StructuredNameOriginal) {
+        StructuredNameOriginal that = (StructuredNameOriginal) other;
+        return this.name.equals(that.name)
+            && this.context.originalName().equals(that.context.originalName());
```

Comment:
Now I'm a bit confused by this. If the context knows both its original name and optimized name, why would we ever not use the optimized name?